### ISSUE TITLE
Sleep wifi usb BCM943602CS (GA-Z77X-D3H)

### ISSUE THREAD

**Issue by domeOo** [theracermaster/Gigabyte-GA-Z77X-macOS-Install#65]
**System information**
 - macOS 10.12.4
 - GA-Z77X macOS Install v2.0.8
 - GA-Z77X-D3H
 - Intel Core i5-XXXXK
 - NVIDIA GeForce GTX 660ti

**Issue description**

Hello,
In connection with the BCM943602CS Wifi Bluetooth combo card sleep isn't working because the USB port where Bluetooth of the card is connect to, isn't marked as internal port (0x255). Maybe u can add a Patch for it.
My Board has 2 internal USB 2.0 Header.

**Comment by theracermaster** [comment]
Hi, can you post a IOReg? What port (look for the ACPI device name in IORegistryExplorer) isn't marked as internal?

**Comment by domeOo** [comment]
![bildschirmfoto 2017-06-20 um 12 47 34](https://user-images.githubusercontent.com/17083036/27329695-d9a4a622-55b6-11e7-8466-1af10e59f57c.png)
![bildschirmfoto 2017-06-20 um 12 48 05](https://user-images.githubusercontent.com/17083036/27329718-f3ca6f78-55b6-11e7-8c4a-92cc4a9f2dbd.png)
![bildschirmfoto 2017-06-20 um 12 48 08](https://user-images.githubusercontent.com/17083036/27329724-f8fa61c4-55b6-11e7-98ac-3045c5a3be73.png)
![bildschirmfoto 2017-06-20 um 12 48 11](https://user-images.githubusercontent.com/17083036/27329726-fbbde110-55b6-11e7-96a9-b4ce32959be2.png)
![bildschirmfoto 2017-06-20 um 12 48 15](https://user-images.githubusercontent.com/17083036/27329748-0db84798-55b7-11e7-98d3-fcfe04addb12.png)
![bildschirmfoto 2017-06-20 um 12 48 20](https://user-images.githubusercontent.com/17083036/27329777-2bcd7618-55b7-11e7-91dc-82d31c7a84e3.png)
![bildschirmfoto 2017-06-20 um 12 48 22](https://user-images.githubusercontent.com/17083036/27329780-2cfcb36e-55b7-11e7-8dcd-2f49afd0b81f.png)

**Comment by zedocrob** [comment]
Hi,
I have the same problem with BCM943602CS (switch with my old TP-LINK for the moment), i've also tried to connect my corsairLink usb (PSU) on internal hub and Sleep is broken too.
Basically with any device connected to this internal hub, sleep is not working.
if you know how to rename it as internal?

10.12.5
GA-Z77X-UD5H
NVIDIA GeForce GTX 660ti

i'v attached my ioreg with USB "corsairlink" connected (not desingned to work on osx but just to show you the problematic port).
it's the same port noticed by domeOo
[zedoc iMac 13,2.ioreg.zip](https://github.com/theracermaster/Gigabyte-GA-Z77X-macOS-Install/files/1099327/zedoc.iMac.13.2.ioreg.zip)

**Comment by domeOo** [comment]
The bluetooth is connected to intern USB 2 Pin header.

**Comment by zedocrob** [comment]
Yes, this one is my front case USB header with a Bluetooth dongle on it.
Because when i use the BCM943602CS BT card on plugged on internal usb header it wakes from sleep like yours.

**Comment by al3xtjames** [comment]
Can you elaborate as to how sleep is broken? Does it instantly wake? If so, can you see Wake reason in Console?

**Comment by zedocrob** [comment]
It instantly wake, in ioreg i see "wake reason: EH01"
![sans-titre](https://user-images.githubusercontent.com/16001729/27791009-59b9a51a-5ff3-11e7-9899-9ec004be9d5b.jpg)

**Comment by domeOo** [comment]
The Computer is going in into sleep but wake up as soon he hits deep sleep

**Comment by zedocrob** [comment]
yes exactly, 1 or 2 seconds after entering deep sleep, it wakes up.

**Comment by domeOo** [comment]
After Hours of investigation with the help of USBinjectAll kext I figured out that the problem is within the Hub driver from Apple. Because the Hub ports Ignore the UsbConnector values. Hint if I Try a port connected to the XHC (Only to front usb 3 connected with internal USB3 Header) where usbConnector is set to 255/0xFF sleep works.

For the GA-Z77X-D3H:

All internal USB 2 Headers are at EH01

All USB 2 Port on Backpanel are at EH02 also the USB 2 ports of the USB3 Connectors on Backpanel are on EH02

The USB 3 Ports of USB 3 Connectors of Backpanel are at XHC

The USB3 Ports of USB  3 internal header are at XHC and the USB2 Ports are also at XHC(These are working with USBConnector value set to 255/0xff 
 
Because the USBConnector flag is ignored the solution may be setting usbType to another value?
So far I found out there is no explanation for all possible values...
I tried 0, 2, and 4 without success.

**Comment by zedocrob** [comment]
I managed to make it sleep on my Z97MX-gaming 5 build:
i routed all usb to XHC removing "FakePCIID_XHCIMux.kext"
see end of this post:
[https://www.tonymacx86.com/threads/guide-creating-a-custom-ssdt-for-usbinjectall-kext.211311/](url)
tested all ports and set internal "HS07" and "HS08" in my case  (internal usb2 ports where BT is plugged in) to 255.
now sleep is working with bluetooth plugged in!
now i will try on my Z77-UD5H when possible.
(i've attached my ssdt for USBinjectAll kext (used on Z97-MX-gaming 5))

[SSDT-USB-ALL.dsl.zip](https://github.com/al3xtjames/Gigabyte-GA-Z77X-macOS-Install/files/1151340/SSDT-USB-ALL.dsl.zip)

**Comment by al3xtjames** [comment]
Most 8/9 chipset boards route all USB ports (including USB 2 ports) to XHC. This isn't possible on 7 series, as USB 2 ports are wired to EHC. Only USB 3 ports are wired to XHC.

**Comment by zedocrob** [comment]
Thanks for the information.
Looks like i will need another BT card for my Z77 build then...

**Comment by domeOo** [comment]
@zedocorb
I use a long USB cable an Connet it via USB 3 hub. So I can set it to 255 and sleep works fine.
But don't forget to deactivate wake from network or else your wifi will be very slow after wake.

**Comment by al3xtjames** [comment]
Could you please try v2.0.9? The internal chipset hubs on EHC1/EHC2 should now be marked as internal. I'm not sure if hubs under that will be counted though.

**Comment by domeOo** [comment]
I will try next week when I'm back at home.
But I also tried this my self last time and it didn't work.
